### PR TITLE
Add list column to null data.table

### LINKED ISSUE #5738: Cannot add list column to empty `data.table`
Both inplace operations currently do not allow adding lists of lengths over 1. The problem arises on both `1.14.8` and current dev.

``` r
library(data.table)
dt = data.table()
x = list("a", 1)
dt[, a := x]
#> Error in `[.data.table`(dt, , `:=`(a, x)): Supplied 2 items to be assigned to 1 items of column 'a'. If you wish to 'recycle' the RHS please use rep() to make this intent clear to readers of your code.
set(dt, j="a", value=x)
#> Error in set(dt, j = "a", value = x): Supplied 2 items to be assigned to 1 items of column 'a'. If you wish to 'recycle' the RHS please use rep() to make this intent clear to readers of your code.
```

Adding columns of other types works fine
``` r
library(data.table)
dt = data.table()
dt[, b := 1:5][]
#>        b
#>    <int>
#> 1:     1
#> 2:     2
#> 3:     3
#> 4:     4
#> 5:     5

dt = data.table()
dt[, c := letters[1:5]][]
#>         c
#>    <char>
#> 1:      a
#> 2:      b
#> 3:      c
#> 4:      d
#> 5:      e
```

### PR DESCRIPTION
**Pull request by ben-schwen**
Closes #5738

- [x] fix
- [x] tests
- [x] NEWS

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -58,6 +58,8 @@
 
 7. `melt` returns an integer column for `variable` when `measure.vars` is a list of length=1, consistent with the documented behavior, [#5209](https://github.com/Rdatatable/data.table/issues/5209). Thanks to @tdhock for reporting and fixing. Any users who were relying on this behavior can change `measure.vars=list("col_name")` (output `variable` was column name, now is column index/integer) to `measure.vars="col_name"` (`variable` still is column name).
 
+8. Adding a list column to an empty `data.table` did not work, [#5738](https://github.com/Rdatatable/data.table/issues/5738). Thanks to Benjamin Schwendinger for the report and the fix.
```
Comment (MichaelChirico) [inline NEWS.md:61]:
```suggestion
8. Adding a list column to an empty `data.table` works consistently with other column types, [#5738](https://github.com/Rdatatable/data.table/issues/5738). Thanks to Benjamin Schwendinger for the report and the fix.
```

File: inst/tests/tests.Rraw [hunk 2]
```diff
@@ -18566,3 +18566,16 @@ test(2261.04, setNumericRounding(2L), 1L)
...
+
+# Add list column to null data.table #5738
+x = list("a", 1)
+dt1 = data.table(a = x)
+dt2 = data.table(a = "a", b=1)
+test(2262.1, null.data.table()[, a := x], dt1)
+test(2262.2, set(null.data.table(), j="a", value=x), dt1)
+test(2262.3, null.data.table()[, c("a","b") := x], dt2)
+test(2262.4, set(null.data.table(), j=c("a","b"), value=x), dt2)
+test(2262.5, null.data.table()[, c("a","b") := list(1:2, 3:4)], data.table(a=1:2, b=3:4))
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:18578]: any reason to use `dt2` above but not e.g. `dt3` here?

File: src/assign.c [hunk 3]
```diff
@@ -360,8 +360,9 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
     // Not possible to test because R won't permit attributes be attached to NULL (which is good and we like); warning from R 3.4.0+ tested by 944.5
   }
   const int nrow = LENGTH(dt) ? length(VECTOR_ELT(dt,0)) :
-                                (isNewList(values) && length(values) ? length(VECTOR_ELT(values,0)) : length(values));
+                                (isNewList(values) && length(values) && (length(values)==length(cols)) ? length(VECTOR_ELT(values,0)) : length(values));
   //                            ^ when null data.table the new nrow becomes the fist column added
+  //                                                              #5738 ^ distinguish between adding list column or multiple columns at once
```
Comment (MichaelChirico) [inline src/assign.c:365]:
```suggestion
  //                                                                    ^ distinguish between adding list column or multiple columns at once, #5738
```

**Comment by MichaelChirico** [comment]
LGTM, superficial fixes only, thanks!

File: inst/tests/tests.Rraw (hunk 2 above)
Comment (ben-schwen) [reply inst/tests/tests.Rraw:18578]: Only the iterative addition of tests 😅

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**